config/routes.rb

```diff
@@ -607,7 +607,8 @@
             put :suspend
             put :resume
           end
-          resources :keys, :controller => 'buyer_application_keys', :only => [:index, :create, :destroy]
+          resources :keys, constraints: { id: %r{([^/]+?)(?=\.json|\.xml|$|/)} }, controller: 'buyer_application_keys', only: %i[index create destroy]
```

Comment:
should we now document that keys ending with `.json` or `.xml` are not supported?

Reply:
They are supported!

Reply:
Ah, so this only overrides the default parsing of them? That's cool. Approved in this case!

Reply:
Yeah, it kind of matches both `.json` and `.xml`, but doesn't include them into the value of the `:id` parameter.

I don't understand the behavior completely, to be honest :sweat_smile: 
When I check this regex with https://regexr.com/, they still get into the same matching group... 
But in Rails it works as expected. For `/keys/abc.def.json`, `/keys/abc.def.xml`, the `:id` parameter is correctly set to `abc.def`, which is what we want.

Reply:
so it means that a key `abc.def.json` is not supported, this is what I meant as a corner case

Reply:
Aaah, I misunderstood then, sorry. I get it now @akostadinov 

Yes, you're right. And indeed, I would consider it to be a corner case. I guess we can't have it all :sweat_smile:

Reply:
What happens if id is just `id: %r{([^/]+?)}` ?

Reply:
@akostadinov If I remember correctly, the value of the `id` param would be `abc.def.json`, so this would not be what we are looking for.

Reply:
@akostadinov we do support both `.xml` and `.json` formats for this endpoint, and the idea was not to change the interface.

Reply:
Actually I think the RE can be improved.

Presently:
```rb
[2] pry(main)> %r{([^/]+?)(?=\.json|\.xml|$|/)}.match "asdasd.json.json"
=> #<MatchData "asdasd" 1:"asdasd">
```

Proposed:
```rb
[6] pry(main)> %r{([^/]+?)(?=(?:\.json|\.xml)?(?:/|$))}.match "asdasd.json.json"
=> #<MatchData "asdasd.json" 1:"asdasd.json">
```

Basically we want only the last `.json` or `.xml` being considered a "format". Any other occurrence should be allowed within the key. 

In this way we can also remove a key ending with `.json` or `.xml` by supplying format to the query. We should do at least in the UI to make sure these edge cases are covered.

Reply:
@akostadinov I don't know, really, the regex in this PR works for me as I explained :shrug: 

I think the router adds something else to this regex (appends something in the beginning and in the end). If you want to open another PR, and test this new regex - be my guest :)

Reply:
presently can a key `asdasd.xml` be handled?

Reply:
> presently can a key `asdasd.xml` be handled?

Yep :+1:

Reply:
how shold query be constructed?

Reply:
All these currently work:

for key `asdf.g`:

```
curl -X 'DELETE' 'http://provider-admin.3scale.localhost:3000/admin/api/accounts/3/applications/686/keys/asdf.g.xml?access_token=admintoken'

curl -X 'DELETE' 'http://provider-admin.3scale.localhost:3000/admin/api/accounts/3/applications/686/keys/asdf.g.json?access_token=admintoken'

curl -X 'DELETE' 'http://provider-admin.3scale.localhost:3000/admin/api/accounts/3/applications/686/keys/asdf.g?access_token=admintoken'
```

for key `asdfg`:

```
curl -X 'DELETE' 'http://provider-admin.3scale.localhost:3000/admin/api/accounts/3/applications/686/keys/asdfg.xml?access_token=admintoken'

curl -X 'DELETE' 'http://provider-admin.3scale.localhost:3000/admin/api/accounts/3/applications/686/keys/asdfg.json?access_token=admintoken' 

curl -X 'DELETE' 'http://provider-admin.3scale.localhost:3000/admin/api/accounts/3/applications/686/keys/asdfg?access_token=admintoken' 
```

Reply:
I make no sense of this RE but it seems to be working. Added a test for it in #3768